// prebuild-install/download.js
/*
The Vent-display project is a ventilator display that consumes PIRDS data and
performs most clinical respiration calculations. This is an important part of
Public Invention's goal of creating an open-source ventilator ecosystem. This
is a stand-alone .html file with about a thousand lines of JavaScript that
implements a clinical display that doctors want to see of an operating
ventilator. It includes live data trace plots of pressure and flow, as well as
calculated values such as tidal volume.
Copyright (C) 2021 Robert Read, Lauria Clarke, Ben Coombs, and Darío Hereñú.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

See the GNU Affero General Public License for more details.
You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

var path = require('path')
var fs = require('fs')
var get = require('simple-get')
var pump = require('pump')
var tfs = require('tar-fs')
var noop = Object.assign({
  http: function () {},
  silly: function () {}
}, require('noop-logger'))
var zlib = require('zlib')
var util = require('./util')
var error = require('./error')
var proxy = require('./proxy')
var mkdirp = require('mkdirp-classic')

function downloadPrebuild (downloadUrl, opts, cb) {
  var cachedPrebuild = util.cachedPrebuild(downloadUrl)
  var tempFile = util.tempFile(cachedPrebuild)
  var log = opts.log || noop

  ensureNpmCacheDir(function (err) {
    if (err) return onerror(err)

    log.info('looking for cached prebuild @', cachedPrebuild)
    fs.access(cachedPrebuild, fs.R_OK | fs.W_OK, function (err) {
      if (!(err && err.code === 'ENOENT')) {
        log.info('found cached prebuild')
        return unpack()
      }

      log.http('request', 'GET ' + downloadUrl)
      var reqOpts = proxy({ url: downloadUrl }, opts)

      if (opts.token) {
        reqOpts.url += '?access_token=' + opts.token
        reqOpts.headers = {
          'User-Agent': 'simple-get',
          Accept: 'application/octet-stream'
        }
      }

      var req = get(reqOpts, function (err, res) {
        if (err) return onerror(err)
        log.http(res.statusCode, downloadUrl)
        if (res.statusCode !== 200) return onerror()
        mkdirp(util.prebuildCache(), function () {
          log.info('downloading to @', tempFile)
          pump(res, fs.createWriteStream(tempFile), function (err) {
            if (err) return onerror(err)
            fs.rename(tempFile, cachedPrebuild, function (err) {
              if (err) return cb(err)
              log.info('renaming to @', cachedPrebuild)
              unpack()
            })
          })
        })
      })

      req.setTimeout(30 * 1000, function () {
        req.abort()
      })
    })

    function onerror (err) {
      fs.unlink(tempFile, function () {
        cb(err || error.noPrebuilts(opts))
      })
    }
  })

  function unpack () {
    var binaryName

    var updateName = opts.updateName || function (entry) {
      if (/\.node$/i.test(entry.name)) binaryName = entry.name
    }

    log.info('unpacking @', cachedPrebuild)

    var options = {
      readable: true,
      writable: true,
      hardlinkAsFilesFallback: true
    }
    var extract = tfs.extract(opts.path, options).on('entry', updateName)

    pump(fs.createReadStream(cachedPrebuild), zlib.createGunzip(), extract,
      function (err) {
        if (err) return cb(err)

        var resolved
        if (binaryName) {
          try {
            resolved = path.resolve(opts.path || '.', binaryName)
          } catch (err) {
            return cb(err)
          }
          log.info('unpack', 'resolved to ' + resolved)

          if (opts.runtime === 'node' && opts.platform === process.platform && opts.abi === process.versions.modules && opts.arch === process.arch) {
            try {
              require(resolved)
            } catch (err) {
              return cb(err)
            }
            log.info('unpack', 'required ' + resolved + ' successfully')
          }
        }

        cb(null, resolved)
      })
  }

  function ensureNpmCacheDir (cb) {
    var cacheFolder = util.npmCache()
    fs.access(cacheFolder, fs.R_OK | fs.W_OK, function (err) {
      if (err && err.code === 'ENOENT') {
        return makeNpmCacheDir()
      }
      cb(err)
    })

    function makeNpmCacheDir () {
      log.info('npm cache directory missing, creating it...')
      mkdirp(cacheFolder, cb)
    }
  }
}

module.exports = downloadPrebuild
